### PR TITLE
MM-22560: Fix crash during slack import

### LINKED ISSUE #13879: Import from slack crashes due to invalid memory access or nil pointer dereference
#### Summary
Import from slack crashes due to invalid memory access or nil pointer dereference

#### Steps to reproduce
1. Created a slack corporate export for last 7 days (only 630KB)
2. Used mattermost cli to import the file as per migrating-from-slack-using-the-mattermost-cli documentation

Mattermost Version: 5.19.0
Running in docker based on https://github.com/mattermost/mattermost-docker

#### Expected behavior
Users, channels, and messages are imported into mattermost, or otherwise ignored if not supported.

#### Observed behavior (that appears unintentional) 
The mattermost cli crashes with a SIGSEGV. The users seem to all be imported; many channels and messages also appear.

The mattermost cli crashes as follow:

```
Running Slack Import. This may take a long time for large teams or teams with many messages.
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x10920bd]

goroutine 1 [running]:
github.com/mattermost/mattermost-server/v5/app.(*App).OldImportChannel(0xc003781500, 0xc00278c680, 0xc0026a8540, 0x9, 0xc0026a8540, 0x9, 0x0, 0x0, 0xc0043ce900, 0x2, ...)
	github.com/mattermost/mattermost-server/v5@/app/slackimport.go:848 +0x41d
github.com/mattermost/mattermost-server/v5/app.(*App).SlackAddChannels(0xc003781500, 0xc00317fb40, 0x1a, 0xc004d34000, 0x390, 0x3bb, 0xc000029998, 0xc0025038f0, 0xc000029968, 0xc005b80b00, ...)
	github.com/mattermost/mattermost-server/v5@/app/slackimport.go:546 +0x76e
github.com/mattermost/mattermost-server/v5/app.(*App).SlackImport(0xc003781500, 0x18f1820, 0xc002525598, 0x9a763, 0xc00317fb40, 0x1a, 0x0, 0x0)
	github.com/mattermost/mattermost-server/v5@/app/slackimport.go:747 +0x1536
github.com/mattermost/mattermost-server/v5/cmd/mattermost/commands.slackImportCmdF(0x241d080, 0xc00026c840, 0x2, 0x2, 0x0, 0x0)
	github.com/mattermost/mattermost-server/v5@/cmd/mattermost/commands/import.go:77 +0x28c
github.com/spf13/cobra.(*Command).execute(0x241d080, 0xc00026c7a0, 0x2, 0x2, 0x241d080, 0xc00026c7a0)
	github.com/spf13/cobra@v0.0.5/command.go:826 +0x460
github.com/spf13/cobra.(*Command).ExecuteC(0x2421400, 0x0, 0x64e00, 0xc00003a118)
	github.com/spf13/cobra@v0.0.5/command.go:914 +0x2fb
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v0.0.5/command.go:864
github.com/mattermost/mattermost-server/v5/cmd/mattermost/commands.Run(...)
	github.com/mattermost/mattermost-server/v5@/cmd/mattermost/commands/root.go:15
main.main()
	github.com/mattermost/mattermost-server/v5@/cmd/mattermost/main.go:30 +0x87
```

#### Possible fixes
The issue is that some users are not supported and are skipped. Then during channel creation, dereferencing Id without checking whether the user exists causes the above crash.

```
diff --git a/app/slackimport.go b/app/slackimport.go
index a8fb13d26..f4fe6ee7c 100644
--- a/app/slackimport.go
+++ b/app/slackimport.go
@@ -846,7 +846,12 @@ func (a *App) OldImportUser(team *model.Team, user *model.User) *model.User {
 
 func (a *App) OldImportChannel(channel *model.Channel, sChannel SlackChannel, users map[string]*model.User) *model.Channel {
        if channel.Type == model.CHANNEL_DIRECT {
-               sc, err := a.createDirectChannel(users[sChannel.Members[0]].Id, users[sChannel.Members[1]].Id)
+               var user1 = users[sChannel.Members[0]]
+               var user2 = users[sChannel.Members[1]]
+               if user1 == nil || user2 == nil {
+                       return nil
+               }
+               sc, err := a.createDirectChannel(user1.Id, user2.Id)
                if err != nil {
                        return nil
                }
@@ -859,7 +864,11 @@ func (a *App) OldImportChannel(channel *model.Channel, sChannel SlackChannel, us
                members := make([]string, len(sChannel.Members))
 
                for i := range sChannel.Members {
-                       members[i] = users[sChannel.Members[i]].Id
+                       var user = users[sChannel.Members[i]]
+                       if user == nil {
+                               return nil
+                       }
+                       members[i] = user.Id
                }
 
                sc, err := a.createGroupChannel(members, users[sChannel.Creator].Id)
```

### PR DESCRIPTION
**Pull request by agnivade**
#### Summary

We were not checking for existence of the users map value
before accessing the Id field. We fix it by checking for
nil pointer first.

Also, we check the length of the Members slice to prevent range
panics.

And finally, while we are here, we move to a switch-case
to make the code a little more idiomatic.

Fixes #13879 

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-22560

### CODE REVIEW

File: app/slackimport.go [hunk 1]
```diff
@@ -845,30 +845,45 @@ func (a *App) oldImportUser(team *model.Team, user *model.User) *model.User {
...
-		sc, err := a.createDirectChannel(users[sChannel.Members[0]].Id, users[sChannel.Members[1]].Id)
+	switch {
+	case channel.Type == model.CHANNEL_DIRECT:
+		if len(sChannel.Members) < 2 {
+			return nil
+		}
+		u1 := users[sChannel.Members[0]]
+		u2 := users[sChannel.Members[1]]
+		if u1 == nil || u2 == nil {
+			return nil
```
Comment (grundleborg) [inline app/slackimport.go]: Should we log an error here instead of silently failing?

File: app/slackimport.go [hunk 2]
```diff
@@ -845,30 +845,45 @@ func (a *App) oldImportUser(team *model.Team, user *model.User) *model.User {
...
 	// check if direct channel has less than 8 members and if not import as private channel instead
-	if channel.Type == model.CHANNEL_GROUP && len(sChannel.Members) < 8 {
+	case channel.Type == model.CHANNEL_GROUP && len(sChannel.Members) < 8:
 		members := make([]string, len(sChannel.Members))
 
 		for i := range sChannel.Members {
-			members[i] = users[sChannel.Members[i]].Id
+			u := users[sChannel.Members[i]]
+			if u == nil {
+				continue
```
Comment (grundleborg) [inline app/slackimport.go]: Also here, maybe log an error. I can imagine the customer support team having a really hard time with complaints from customers when the slack importer appears to randomly miss out data without helpful errors.

Comment (agnivade) [reply app/slackimport.go]: done.

**Comment by agnivade** [comment]
Why u no merge mattermod ?

**Pull request merged**